### PR TITLE
1705 BE Daniel Vogelsang Independent Timed Assessment

### PR DESCRIPTION
**Pull request by thedanielvogelsang**
@mikedao @laurenfazah

### CODE REVIEW

File: db/migrate/20170801160824_create_taggings.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class CreateTaggings < ActiveRecord::Migration[5.0]
+  def change
+    create_table :taggings do |t|
+      t.references :tag, foreign_key: true
+      t.references :job, foreign_key: true
+
```
Comment (laurenfazah) [inline db/migrate/20170801160824_create_taggings.rb:6]: I would have loved to see a test drive the creation of this join table.

File: spec/features/jobs/user_visits_job_page_and_sees_tags_spec.rb [hunk 2]
```diff
@@ -7,8 +7,13 @@
...
     job = Job.create(title: "JobTitle", description: "tough job", level_of_interest: 20, company_id: company.id, city: "Denver", salary: 20000)
+    job.tags << tag1
+    job.tags << tag3
 
     visit(company_job_path(company, job))
     save_and_open_page
+    expect(page).to have_content(tag1.name)
+    expect(page).to have_content(tag3.name)
+    expect(page).to_not have_content(tag2.name)
   end
```
Comment (laurenfazah) [inline spec/features/jobs/user_visits_job_page_and_sees_tags_spec.rb]: Good use of an edge case test here with making sure tag2's content is not showing.

File: app/models/tag.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 class Tag < ApplicationRecord
   has_many :taggings
   has_many :jobs, through: :taggings
+
+  def job_count
+    self.jobs.count
+  end
```
Comment (laurenfazah) [inline app/models/tag.rb]: Great helper method. Would love to see it unit test driven, though.

File: spec/models/tag_spec.rb [hunk 4]
```diff
@@ -1,5 +1,17 @@
...
+      company = Company.create(name: "CompanyTitle")
+      job = Job.create(title: "JobTitle", description: "tough job", level_of_interest: 20, company_id: company.id, city: "Denver", salary: 20000)
+      job.tags << tag1
+      job2 = Job.create(title: "SecondJob", description: "tough job", level_of_interest: 20, company_id: company.id, city: "Denver", salary: 20000)
+      job2.tags << tag1
+      job3 = Job.create(title: "ThirdJob", description: "tough job", level_of_interest: 20, company_id: company.id, city: "Denver", salary: 20000)
+      job3.tags << tag1
+      expect(tag1.job_count).to eq(3)
+    end
+  end
```
Comment (laurenfazah) [inline spec/models/tag_spec.rb]: Great set up. Edge case testing with additional tags created could make it even stronger.

File: app/models/tag.rb [hunk 5]
```diff
@@ -5,4 +5,8 @@ class Tag < ApplicationRecord
   def job_count
     self.jobs.count
   end
+
+  def average_salary
+    self.jobs.average("salary")
+  end
```
Comment (laurenfazah) [inline app/models/tag.rb]: Same as with `job_count` - would love to see a unit test behind this method.

**Comment by laurenfazah** [comment]
Strong solution, Daniel. I think your feature test setup was really strong throughout and clearly kept you on track towards those three solutions.

The only comments I have to make are:

1) more edge case / sad path testing could be added to tests to make them even stronger
2) stick to TDD as much as possible - I saw this slip with the join table creation and missed with the Tag instance methods.

Other than that, refactors within your tests are great as is the solution you reached.
